### PR TITLE
Fix WebSite projects opened from IIS

### LINKED ISSUE #12337: WebSite projects opened from IIS fail to install packages
### NuGet Product Used

Visual Studio Package Management UI, Visual Studio Package Manager Console

### Product Version

VS 17.4

### Worked before?

17.3

### Impact

I'm unable to use this version

### Repro Steps & Context

1. In IIS Manager, create a virtual directory as an application.
2. In VS, use "open website", select "local IIS", and then select the directory
3. Try to install a package that has `lib/<tfm>/*.dll` files.

An obscure error message is reported: 

> Failed to add reference to ‘Newtonsoft.Json’.
> 
> Invalid URI: The Authority/Host could not be parsed.

### Verbose Logs

_No response_

### PR DESCRIPTION
**Pull request by zivkan**
IPersistFileFormat.GetCurFile was strangely returning the IIS http(s) URL, not the local folder name.
GetCanonicalName is returning the path, with trailing \, so that Path.GetDirectory will still return the same path

## Bug

Fixes: https://github.com/NuGet/Home/issues/12337

Regression? **yes**
Last working version: 6.3.x

## Description

https://github.com/NuGet/NuGet.Client/pull/4698 was aimed at reducing usage of `DTE`, but it also changed `VsHierarchyUtility.GetProjectPath` from using `IVsHierarchy.GetCanonicalName` to `IPersistFileFormat.GetCurFile`.

However, when a WebSite project is loaded from the UI to select a local IIS directory, rather than loading a solution file that already has the path to the folder, `IPersistFileFormat` will strangely return the `http(s)` URL, not the filesystem path, to the project/website folder.

Given the previous PR was aimed at reducing `DTE` usage, and `IVsHierarchy` is not `DTE`, it's not clear to me why this was changed. It's not described in the PR either, so I don't see any reason to assume there's a problem changing it back.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [x] **Test exception**: Requires IIS to be installed on the test machine, IIS must have a web application configured, and the website project to be loaded from the IIS configuration, not a solution file or new project.
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: src/NuGet.Clients/NuGet.VisualStudio.Common/IDE/VsHierarchyUtility.cs [hunk 1]
```diff
@@ -25,8 +25,7 @@ public static class VsHierarchyUtility
         public static string GetProjectPath(IVsHierarchy project)
         {
             ThreadHelper.ThrowIfNotOnUIThread();
-            var format = (IPersistFileFormat)project;
-            format.GetCurFile(out string projectPath, out uint _);
+            project.GetCanonicalName((uint)VSConstants.VSITEMID.Root, out string projectPath);
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/IDE/VsHierarchyUtility.cs:28]: @nkolev92 any chance you remember why you changed this in your PR?  It's unrelated to DTE, so I don't think it has any of the perf/other issues of using DTE.

**Review by donnie-msft — APPROVED** [review]
Isn't it kind of a red flag that no test exists that attempted to install packages to a Website project? 
Edit: Oh right, we have EndToEnd tests for Websites, but not using DTE/VS. Hmm..

**Comment by zivkan** [comment]
> Oh right, we have EndToEnd tests for Websites, but not using DTE/VS. Hmm..

I'm not sure what you mean by "not using DTE/VS".  But as I wrote in the PR checklist's test exception section, it's difficult to test because this bug happens only when the WebSite project is loaded from IIS. It doesn't happen when the WebSite project is loaded from the filesystem. Setting this up programmatically is non-trivial (assuming VS even has an API to programmatically "open project from local IIS URL"). Our test vendors have a manual test for this now, so they'll periodically/regularly check for regressions.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/IDE/VsHierarchyUtility.cs (hunk 1 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/IDE/VsHierarchyUtility.cs:28]: I'd reach out to David about this.
I changed it with his guidance so he'll likely know better.
I think it was not returning the correct value for some type of projects. It was null but we needed a non null value.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/IDE/VsHierarchyUtility.cs:28]: It was a special project type such like maybe shared project or c++/CLI

Comment (zivkan) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/IDE/VsHierarchyUtility.cs:28]: I should have posted this a few days ago, but when debugging I found that for native C++ and C++/CLI projects, both APIs return the same value (the full path, including the `vcxproj` file. However, for .NET projects, `IVsHierarchy.GetCanonicalName` returns only the directory, not the project file, whereas `IPersistFileFormat.GetCurFile` does return the project file. This gets passed to the our project adapter in a parameter whose name expects the project filename.  I didn't do much manual testing, and I assumed that E2E test failures were flakiness, not regressions.

So, this needs to be investigated, but assuming it's flakiness, it's not clear to me how important it is that we get this working  the same. If we need it to keep returning the project file for C# projects (a concept that doesn't even exist for WebSite projects), then our `VsProjectAdapter` class can no longer be project type agnostic. We'll need it to intrinsically understand different project types and how different ones need to use different APIs. And doing that will be a much bigger change.

**Comment by zivkan** [comment]
I spent half a day on this today, trying to change NuGet's assumption that all `VsProjectAdapter`s can work the same, and try to make it part of the project system specific abstractions.   It's really not easy 😞. This is going to take quite a bit of extra work to get working properly.

Switching to draft mode, although I'm not sure I'll have time to work on it this sprint, before the bot closes it as stale.

**Comment by PaavoNieminen** [comment]
Hope this dos not get buried as the pain caused by this bug is somewhat annoying ...

**Comment by zivkan** [comment]
Sorry, there's been a lot going on, so I haven't made the time to work on this. The PR that introduced the regression is a perf improvement, so at this point I think it's easiest to revert the whole commit. We'll need to find a better solution to get the perf win back again.

**Pull request closed without merging**